language/move-prover/spec-lang/src/translate.rs

```diff
@@ -1609,6 +1609,73 @@ impl<'env, 'translator, 'module_translator> ExpTranslator<'env, 'translator, 'mo
 /// ## Type Translation
 
 impl<'env, 'translator, 'module_translator> ExpTranslator<'env, 'translator, 'module_translator> {
+    /// Translates an hlir type into a target AST type.
+    fn translate_hlir_single_type(&mut self, ty: &SingleType) -> Type {
+        use move_lang::hlir::ast::SingleType_::*;
+        match &ty.value {
+            Ref(is_mut, ty) => {
+                Type::Reference(*is_mut, Box::new(self.translate_hlir_base_type(&*ty)))
+            }
+            Base(ty) => self.translate_hlir_base_type(&*ty),
+        }
+    }
+
+    fn translate_hlir_base_type(&mut self, ty: &BaseType) -> Type {
+        use move_lang::{
+            hlir::ast::{BaseType_::*, TypeName_::*},
+            naming::ast::BuiltinTypeName_::*,
+        };
+        match &ty.value {
+            Param(TParam { debug, .. }) => {
+                let sym = self.symbol_pool().make(debug.value.as_str());
+                self.type_params_table[&sym].clone()
+            }
+            Apply(_, type_name, args) => {
+                let loc = self.to_loc(&type_name.loc);
+                let arg_types = self.translate_hlir_base_types(args);
+                match &type_name.value {
+                    Builtin(builtin_type_name) => match &builtin_type_name.value {
+                        Address => Type::new_prim(PrimitiveType::Address),
+                        U8 => Type::new_prim(PrimitiveType::U8),
+                        U64 => Type::new_prim(PrimitiveType::U64),
+                        U128 => Type::new_prim(PrimitiveType::U128),
+                        Vector => Type::Vector(Box::new(self.translate_hlir_base_type(&args[0]))),
+                        Bool => Type::new_prim(PrimitiveType::Bool),
+                    },
+                    ModuleType(m, n) => {
+                        let module_name = ModuleName::from_str(
+                            &m.0.value.address.to_string(),
+                            self.symbol_pool().make(m.0.value.name.0.value.as_str()),
+                        );
+                        let symbol = self.symbol_pool().make(n.0.value.as_str());
+                        let qsym = QualifiedSymbol {
+                            module_name,
+                            symbol,
+                        };
+                        let rty = self.parent.parent.lookup_type(&loc, &qsym);
+                        if !args.is_empty() {
+                            // Replace type instantiation.
+                            if let Type::Struct(mid, sid, _) = &rty {
+                                Type::Struct(*mid, *sid, arg_types)
+                            } else {
+                                unreachable!()
```

Comment:
rty is not guaranteed to be a struct AFAICT, because lookup_type can fail (I know it shouldn't because move-lang is type checked, but never know whether there is a bug). I guess you need to check whether its the Error type, and propagate it in this case.

Reply:
Done